src/ray/raylet/placement_group_resource_manager.cc

```diff
@@ -136,16 +148,103 @@ void OldPlacementGroupResourceManager::ReturnBundle(
+      // return bundle resources.
+      ReturnBundle(bundle_spec);
+    }
+  }
+
+  std::shared_ptr<TaskResourceInstances> resource_instances =
+      std::make_shared<TaskResourceInstances>();
+  bool allocated = cluster_resource_scheduler_->AllocateLocalTaskResources(
+      bundle_spec.GetRequiredResources().GetResourceMap(), resource_instances);
+
+  if (!allocated) {
+    return false;
+  }
+
+  auto bundle_state =
+      std::make_shared<BundleTransactionState>(CommitState::PREPARED, resource_instances);
+  pg_bundles_[bundle_spec.BundleId()] = bundle_state;
+  bundle_spec_map_.emplace(bundle_spec.BundleId(), std::make_shared<BundleSpecification>(
+                                                       bundle_spec.GetMessage()));
+
+  return true;
+}
+
+void NewPlacementGroupResourceManager::CommitBundle(
+    const BundleSpecification &bundle_spec) {
+  auto it = pg_bundles_.find(bundle_spec.BundleId());
+  if (it == pg_bundles_.end()) {
+    // We should only ever receive a commit for a non-existent placement group when a
+    // placement group is created and removed in quick succession.
+    RAY_LOG(DEBUG)
+        << "Received a commit message for an unknown bundle. The bundle info is "
+        << bundle_spec.DebugString();
+    return;
+  } else {
+    // Ignore request If the bundle state is already committed.
+    if (it->second->state_ == CommitState::COMMITTED) {
+      RAY_LOG(INFO) << "Duplicate committ bundle request, skip it directly.";
+      return;
+    }
+  }
+
+  const auto &bundle_state = it->second;
+  bundle_state->state_ = CommitState::COMMITTED;
+
+  for (const auto &resource : bundle_spec.GetFormattedResources()) {
+    cluster_resource_scheduler_->AddLocalResource(resource.first, resource.second);
+  }
+}
+
+void NewPlacementGroupResourceManager::ReturnBundle(
+    const BundleSpecification &bundle_spec) {
+  auto it = pg_bundles_.find(bundle_spec.BundleId());
+  if (it == pg_bundles_.end()) {
+    RAY_LOG(INFO) << "Duplicate cancel request, skip it directly.";
+    return;
+  }
+  const auto &bundle_state = it->second;
+  if (bundle_state->state_ == CommitState::PREPARED) {
+    CommitBundle(bundle_spec);
+  }
+
+  // Return original resources to resource allocator `ClusterResourceScheduler`.
+  auto original_resources = it->second->resources_;
+  cluster_resource_scheduler_->FreeLocalTaskResources(original_resources);
+
+  // Substract placement group resources from resource allocator
+  // `ClusterResourceScheduler`.
+  const auto &placement_group_resources = bundle_spec.GetFormattedResources();
+  std::shared_ptr<TaskResourceInstances> resource_instances =
+      std::make_shared<TaskResourceInstances>();
+  cluster_resource_scheduler_->AllocateLocalTaskResources(placement_group_resources,
+                                                          resource_instances);
+  for (const auto &resource : placement_group_resources) {
+    if (cluster_resource_scheduler_->IsAvailableResourceEmpty(resource.first)) {
```

Comment:
When is the case it is not empty? Can we just call RAY_CHECK instead of if?

Reply:
no, we can't, because wildcard resource name like 'CPU_group_zzz' maybe coincide in two bundles.